File: 19. 删除链表的倒数第 N 个结点.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */

 var removeNthFromEnd = function(head, n) {
    var st = new ListNode(0, head);
    var left = st;
    var right = head;
    for (var i = 0; i < n; i++) right = right.next;
    while (right) {
        left = left.next;
        right = right.next;
    }
    left.next = left.next.next;
    return st.next;
};

File: 19. 删除链表的倒数第 N 个结点.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummyHead = new ListNode();
        dummyHead->next = head;

        ListNode* first = dummyHead;
        ListNode* second = dummyHead;

        while (n--) first = first->next;
        while (first->next) {
            first = first->next;
            second = second->next;
        }
        second->next = second->next->next;

        return dummyHead->next;
    }
};